Draft initial essential stuff

BehrRiley opened the pull request:
Adds various commands and a command dependency proposal;

**Command summary**:
`/fly (player)` | For toggling another player's or your own flight
`/gmc (player)` | For setting another player's or your own gamemode to creative
`/gms (player)` | For setting another player's or your own gamemode to survival
`/gmsp (player)` | For setting another player's or your own gamemode to spectator
`/gma (player)` | For setting another player's or your own gamemode to adventure
`/heal (player)` | For healing another player or yourself
`/suicide` | For killing yourself; generally for like, if you get stuck someplace but we also have teleport commands
`/teleport help / (player) ((x y z (world)) / (world) / (player))` | For general teleporting such as:
- to coordinates in the world you're in
- to coordinates in another world
- to another world's spawn location
- to another player's location
additionally the option of teleporting a specified player to all of the cases above
`/time <Time of Day/0-23999>` | For changing the time of the day in a world to either a tick-based time of the day, or any of the common named times of the day; included "bedtime" for exact timing of when Villagers go to bed as well as other villager-relevant time-of-day names, and "sunset" for when mobs start to spawn at night
`/weather <sunny/storm/thunder/rainy> (world)` | For setting the weather in optionally a specified world; added "rainy" as a more familiar option to thunder/storming

BehrRiley committed:
Draft initial essential stuff

Adds various commands and a command dependency proposal

BehrRiley committed:
Remove fly command

Removed per greenleew:
```xml
<greenleeuw — Today at 12:17 PM>
Also /fly is connected with our vote prizes including fly time
```

acikek commented on essentials/command_dependencies.dsc:
```diff
@@ -0,0 +1,114 @@
+  script:
+  - if <[user].length> < 3:
+    - inject player_verification_offline.too_short
+  - else if <server.match_player[<[user]>].if_null[invalid]> == invalid:
```
`.exists`

acikek commented on essentials/command_dependencies.dsc:
```diff
@@ -0,0 +1,114 @@
+    type: task
+    debug: false
+    script:
+        - define command "<queue.script.data_key[aliases].first.if_null[<queue.script.data_key[name]>]> "
```
should probably keep indentation consistent (throughout files at least)

BehrRiley replied:
resolved in a queued commit

acikek commented on essentials/command_dependencies.dsc:
```diff
@@ -0,0 +1,114 @@
+  debug: false
+  definitions: user
+  too_short:
+    - define hover "<proc[colorize].context[You typed<&co>]><n><&color[#ff3333]>/<context.command.to_lowercase> <context.raw_args>"
```
this line is used quite a lot, maybe make it a proc of its own?

BehrRiley replied:
it's used four times but single-line procedure scripts are kind of bloaty

acikek commented on essentials/gamemode.dsc:
```diff
@@ -0,0 +1,81 @@
+    1: <server.online_players.parse[name].exclude[<player.name>]>
+  script:
+    - define gamemode survival
+    - inject gamemode_task
```
its unnecessary to inject `gamemode_task` into these 4 commands

acikek commented on essentials/gamemode.dsc:
```diff
@@ -0,0 +1,81 @@
+        - narrate "<[user].name> <&color[#ff3333]>is already in <[gamemode]> mode."
+      - else:
+        - if <[user].name> != <player.name>:
+          - narrate "<[user].name> <&color[#33ff33]>'s gamemode changed to: <&color[#f4ffb3]><[gamemode]>"
```
no else case here, both will narrate if true (intentional?)

BehrRiley replied:
This is correct; if a player named `Behr_Riley` were to type `/gmc behr_riley`, instead of him receiving both messages, because his name equals the name he specified, he will only be narrated the message that his gamemode has changed

acikek replied:
still unclear if this is intentional or not. there is no else case, so if this if passes, meaning the player name isnt the name, then both will narrate

BehrRiley replied:
resolved via discord converstaion; just differentiates messages based on:
- if it's a player toggling their own gamemode
- if it's a player toggling another player's gamemode
- if it's a player toggling another player's gamemode but that other player is actually themself

acikek commented on essentials/gamemode.dsc:
```diff
@@ -0,0 +1,81 @@
+      - else:
+        - if <[user].name> != <player.name>:
+          - narrate "<[user].name> <&color[#33ff33]>'s gamemode changed to: <&color[#f4ffb3]><[gamemode]>"
+        - narrate targets:<[user]> "<&color[#33ff33]>Gamemode changed to: <&color[#f4ffb3]><[gamemode]>"
```
move this logic to the args empty check (line 63), and add an or case for the name check

acikek commented on essentials/gamemode.dsc:
```diff
@@ -0,0 +1,81 @@
+    - define gamemode survival
+    - inject gamemode_task
+
+gamemode_task:
```
tab completions? see `heal_command`

BehrRiley replied:
This is a task script; the gamemode commands all use tab completions though yeah

acikek replied:
not entirely sure what I was referencing here

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+  aliases:
+    - tp
+  permission: behrry.essentials.teleport
+  tab complete:
```
i recommend zozer's TabCompleteEngine for this mess

BehrRiley replied:
I do not, his is very procedural and this is just linear based on how many arguments are being used

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+  name: teleport
+  debug: false
+  description: Teleports you or another player to a location, to a world, or to another player
+  usage: /teleport help / (player) ((x y z (world)) / (world) / (player))
```
`/teleport [help] / [<&gt>player<&lt>] (<&gt>x<&lt> <&gt>y<&lt> <&gt>z<&lt>) [<&gt>world<&lt>/<&gt>player<&lt>]`
equivalent to
`/teleport [help] / [<player>] (<x> <y> <z>) [<world>/<player>]`

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+      - determine <server.worlds.parse[name].filter[starts_with[<context.args.get[5].if_null[<empty>]>]]>
+
+  help:
+    - narrate "<&6>/<&e>teleport x y z <&b>| <&a>Teleports you to the coordinates <&2><&lt><&a>x y z<&2><&gt>"
```
make a list in a data script and loop over this

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+        - inject command_error
+
+      - foreach <[coordinates]> as:coordinate:
+        - if <[coordinate].abs> > 9000:
```
should this be 8000?

BehrRiley replied:
I'll be honest and confess i both:
- intentionally lied to my players here to ensure they didn't get too close to the world border i hard-coded
- am unsure of what the real world border is lol

acikek replied:
take advantage of `WorldTag.border_size`

BehrRiley replied:
This tag would be invalid as the server doesn't actually use the world border; it has some form of event handling for when a player exceeds the allowed limits

we do how-ever need clarification on the world border

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+          - define grammar_noun coordinates
+        - else:
+          - define grammar_verb "is an"
+          - define grammar_noun coordinate
```
repeated reason define would be shorter

BehrRiley replied:
?

acikek replied:
defining this grammar stuff is unnecessary and convoluted. just check if the list's size is > 1 and then use a define command for each case

BehrRiley replied:
oh, i see what you mean now

acikek commented on essentials/teleport.dsc:
```diff
@@ -0,0 +1,210 @@
+        - if <[coordinate].abs> > 9000:
+          - define reason "Exceeds the server's border: 8000 to -8000 across the X and Z axis"
+          - inject command_error
+  script:
```
i dont even want to look at this script

acikek commented on essentials/time.dsc:
```diff
@@ -0,0 +1,63 @@
+  # % ██ [ Match time with time of day by name ] ██
+    - else:
+      - define time_of_day <context.args.first>
+      - choose <[time_of_day]>:
```
is `choose` case insensitive? if not, may want to lowercase this

BehrRiley replied:
no

acikek commented on essentials/weather.dsc:
```diff
@@ -0,0 +1,42 @@
+  # % ██  [ Check if a player specified a world ]  ██
+    - if <context.args.size> == 2:
+      - define world <context.args.last>
+      - if !<Server.worlds.parse[name].contains[<[world]>]>:
```
`server`

BehrRiley committed:
description tweak and convenience tag update

- minor tweak to `/heal` command description
- utilized `.exists` subtag in favor of `.if_null[invalid] == invalid` tags

BehrRiley committed:
Added more commands and minor tweaks

- Adds `/top` command to teleport a player to the highest solid block at their location
- Adds /maxhealth` command to adjust a player's maximum health
- Minor update to the `weather` command script that had randomly capitalized tag names
- Moved duplicate code in the teleport command to sub-paths
- Added verbose tab-completion commentary to teleport command
- Changed a hacky location dissection used to grab the player's relative (~) coordinates

BehrRiley committed:
Moved script command dependencies to utilities folder

BehrRiley committed:
Adds multiple commands, moves command dependency

- Move command dependencies to utilities folder since, well, it's a utility to scripts
- Add `/feed (player (#))` to satiate or intensify a player's hunger
- Add `/oxygen (player) <#>` to set another player's or your own oxygen level
- Add `/clear_inventory (player)` to clear another player's or your own inventory

BehrRiley commented:
- Moved command dependencies to utilities folder since, well, it's a utility to scripts
- Tweaked some spacing I forgot to adjust for consistency
- Add `/feed (player (#))` to satiate or intensify a player's hunger
- Add `/oxygen (player) <#>` to set another player's or your own oxygen level
- Add `/clear_inventory (player)` to clear another player's or your own inventory

acikek commented on essentials/clear_inventory.dsc:
```diff
@@ -0,0 +1,29 @@
+      - if <[user].name> != <player.name>:
+        - narrate "<[user].name><&color[#33ff33]>'s inventory was cleared"
+      - narrate targets:<[user]> "<&color[#33ff33]>Your inventory was cleared"
+      - inventory clear d:<[user].inventory>
```
I feel like this logic is used a lot and may be better as an injectable task

acikek commented on essentials/feed.dsc:
```diff
@@ -0,0 +1,60 @@
+        - narrate targets:<[user]> "<&color[#33ff33]>Your hunger was satiated."
+      - else:
+        - if <[user].name> != <player.name>:
+          - narrate "<[user].name>'s hunger was increased."
```
`was intensified`

BehrRiley committed:
narrate usage tweak

acikek commented:
Does `clear_inventory` even need to be a command? The default `/clear` has the exact same function and is already builtin

acikek committed:
clean tp help
